lib/swoosh/adapters/amazon_ses.ex

```diff
@@ -0,0 +1,229 @@
+      |> generate_signing_string(config, current_date_time)
+      |> generate_signature(current_date_time, config[:region], config[:secret])
+
+    authorization = prepare_authorization(config, signed_header_list, current_date_time, signature)
+    Map.put(headers, "Authorization", authorization)
+  end
+
+  defp setup_headers_string(headers) do
+    headers
+    |> Map.keys
+    |> Enum.sort
+    |> Enum.map(fn(key) -> "#{String.downcase(key)}:#{headers[key]}" end)
+    |> Enum.join("\n")
+  end
+
+  defp generate_signed_header_list(headers) do
+    headers
+    |> Map.keys
+    |> Enum.map(fn(key) -> String.downcase(key) end)
+    |> Enum.join(";")
+  end
+
+  defp determine_request_hash(query, headers, signed_header_list) do
+    hashed_query =
+      :crypto.hash(:sha256, query)
+      |> Base.encode16
+      |> String.downcase
+
+    canonical_request =
+      [
+        "POST",
+        "/",
+        "",
+        "#{headers}",
+        "",
+        signed_header_list,
+        hashed_query
+      ]
+      |> Enum.join("\n")
+
+    :crypto.hash(:sha256, canonical_request) |> Base.encode16 |> String.downcase
+  end
+
+  defp prepare_header_host(headers, config) do
+    Map.put(headers, "Host", @host_prefix <> config[:region] <> @host_suffix)
+  end
+
+  defp prepare_header_date(headers, date_time) do
+    Map.put(headers, "X-Amz-Date", normalize_iso8601(date_time))
+  end
+
+  defp prepare_header_length(headers, query) do
+    Map.put(headers, "Content-Length", String.length(query))
+  end
+
+  defp prepare_authorization(config, signed_header_list, date_time, signature) do
+    date = extract_date_string(date_time)
+    credential = "#{config[:access_key]}/#{date}/#{config[:region]}/#{@service_name}/aws4_request"
+    "#{@encoding} Credential=#{credential}, SignedHeaders=#{signed_header_list}, Signature=#{signature}"
+  end
+
+  defp generate_signature(string_to_sign, date_time, region, secret) do
+    "AWS4" <> secret
+    |> encrypt_value(extract_date_string(date_time))
+    |> encrypt_value(region)
+    |> encrypt_value(@service_name)
+    |> encrypt_value("aws4_request")
+    |> encrypt_value(string_to_sign)
+    |> Base.encode16
+    |> String.downcase
+  end
+
+  defp generate_signing_string(request_hash, config, dt) do
+    date = extract_date_string(dt)
+    normalized_date_time = normalize_iso8601(dt)
+
+    [
+      @encoding,
+      "#{normalized_date_time}",
+      "#{date}/#{config[:region]}/#{@service_name}/aws4_request",
+      request_hash
+    ]
+    |> Enum.join("\n")
+  end
+
+  defp encrypt_value(secret, unencrypted_data), do: :crypto.hmac(:sha256, secret, unencrypted_data)
+
+  defp extract_date_string(dt), do: "#{dt.year}#{dt.month}#{dt.day}"
+  defp normalize_iso8601(dt) do
+    date = extract_date_string(dt)
+    {_, time} = Time.new(dt.hour, dt.minute, dt.second)
+    time_string =
+      time
+      |> Time.to_string 
+      |> String.replace(":", "")
```

Comment:
You did `"#{dt.year}#{dt.month}#{dt.day}"` for date, why not just `"#{dt.hour}#{dt.minute}#{dt.second}"` here as well? `dt` is generated by `DateTime.utc_now` anyway, it couldn't be a bad value.

Reply:
I believe the reason I went about creating the time object was that the to_string value kept the 0 padding at the beginning of single digit fields. So it came out to 02:10:20 with the Time.to_string and replace vs 2:10:20 with the `#{dt.hour}#{dt.minute}#{dt.second}`. If I remember right, Amazon was yelling at me when I tried to send up a datetime would that zero padding at the beginning.

Reply:
Tried a little something different here. Used the `String.pad_leading` function instead of making that new time object and going that route.